algebra/src/fft/domain/basic_radix_2_domain.rs

```diff
@@ -0,0 +1,286 @@
+use crate::{FpParameters, PrimeField};
+use crate::{multicore::Worker, EvaluationDomainImpl};
+use std::fmt;
+use rayon::prelude::*;
+
+/// Defines a domain over which finite field (I)FFTs can be performed. Works
+/// only for fields that have a large multiplicative subgroup of size that is
+/// a power-of-2. The roots of the target polynomial will be the powers of a
+/// 2^n root of unity in the field. This allows us to perform polynomial operations
+/// in O(n) by performing an O(n log n) FFT over such a domain.
+#[derive(Copy, Clone, Hash, Eq, PartialEq, Default)]
+pub struct BasicRadix2Domain<F: PrimeField> {
+    /// The size of the domain.
+    pub size:                  u64,
+    /// `log_2(self.size)`.
+    pub log_size_of_group:     u32,
+    /// Size of the domain as a field element.
+    pub size_as_field_element: F,
+    /// Inverse of the size in the field.
+    pub size_inv:              F,
+    /// A generator of the subgroup.
+    pub group_gen:             F,
+    /// Inverse of the generator of the subgroup.
+    pub group_gen_inv:         F,
+    /// Multiplicative generator of the finite field.
+    pub generator_inv:         F,
+}
+
+impl<F: PrimeField> fmt::Debug for BasicRadix2Domain<F> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "Multiplicative subgroup of size {}", self.size)
+    }
+}
+
+impl<F: PrimeField> BasicRadix2Domain<F> {
+
+    /// Given an index which assumes the first elements of this domain are the elements of
+    /// another (sub)domain with size size_s,
+    /// this returns the actual index into this domain.
+    pub fn reindex_by_subdomain(&self, other: Self, index: usize) -> usize {
```

Comment:
This function is used by Marlin to map the first R1CS variables, which correspond to the public inputs, to the input domain S as subgroup of H.
In my opinion a more elegant way is to use the decomposition via the group action of S on H as done for the FFT, i.e.
x =  i |S| + (x mod |S|)= i |S| + j  -> i +  j * |H|/|S|.